### PR TITLE
Add 'w' and 's' bit to xarch instruction flags.

### PR DESCRIPTION
**Pull request by anthonycanino**
This PR addresses https://github.com/dotnet/runtime/issues/35305.

## Changes

Change encodes 'w' and 's' in the insFlags struct and INS_FLAG entry for xarch instruction table. In addition, `HasWBit` and `HasSBit` check if this flag is set for an instruction, which allows to start simplifying some of the various ad-hoc checks for these bits that were previously done per-instruction throughout emitxarch.cpp.

## Discussion

Please see https://github.com/dotnet/runtime/issues/35305 for an extended discussion on adding the 'w' and 's' bit as INS_FLAGS entries. 

There is one point that I need some expert knowledge on: whether or not we need `isInsCMOV` and `!isInsCMOV` at two different points. I left comments in the code to draw attention to it: https://github.com/dotnet/runtime/commit/33a9ddbec24557c78d0feb8f80ad5ed0649b349d#diff-6b4e0f32449f2f144e05699f59f74415a564693637f643084a896dfbd081830dR10984 and https://github.com/dotnet/runtime/commit/33a9ddbec24557c78d0feb8f80ad5ed0649b349d#diff-6b4e0f32449f2f144e05699f59f74415a564693637f643084a896dfbd081830dR11449

### CODE REVIEW

**Comment by JulieLeeMSFT** [comment]
@tannergooding PTAL the community PR.

**Comment by anthonycanino** [comment]
I believe the Linux x64 test Interop/PInvoke/Generics/GenericsTest/GenericsTest.sh currently failing on main (https://github.com/dotnet/runtime/issues/61300)

File: src/coreclr/jit/emitxarch.cpp
Comment (tannergooding) [inline src/coreclr/jit/emitxarch.cpp:166]:
```suggestion
// HasWBit: check if the instruction has the 'w' bit required to use the
```

**Comment by tannergooding** [comment]
The changes generally LGTM. I still need to finish going through the instruction tables to validate the `instrsxarch.h` changes all look good.

File: src/coreclr/jit/emitxarch.cpp
Comment (tannergooding) [inline src/coreclr/jit/emitxarch.cpp:160]: For `s` in particular, I wonder if there needs to be a specialization of it...

For example, if we look at 
image

We'll see that all variants have the `w` bit; but only 2 variants has the `s` bit (and have the same pattern). There is an alternative/shorter encoding that does `immediate to AL, AX, or EAX` which doesn't use the `s` bit.

Likewise if we look at the 64-bit mode variants:
image

There are similar cases for the w/s-bit and so "just the flag" doesn't look sufficient to know "does the bit exist and is it valid to set"

Comment (tannergooding) [reply src/coreclr/jit/emitxarch.cpp:160]: The annotations otherwise all look correct for where some encoding for an instruction uses the `s` and or `w` bits, respectively.

Comment (anthonycanino) [reply src/coreclr/jit/emitxarch.cpp:160]: So do you think that we need a combination of looking at what form the instruction is in and whether it has the w/s bit to properly implement `HasSBit` and `HasWBit`?

Comment (tannergooding) [reply src/coreclr/jit/emitxarch.cpp:160]: That's somewhat the concern, yes.

Basically, simply annotating `Has_Sbit` and `Has_Wbit` isn't enough and might cause later downstream bugs. It really is a few factors that determine if the S/W bit are present and if it means anything.

Comment (anthonycanino) [reply src/coreclr/jit/emitxarch.cpp:160]: Ok, I understand.

I have to go back over the tables to understand how many cases there are that might actually arise, or if there is a general enough pattern that most of the logic can be coded in using the `Has_Sbit` and `Has_Wbit` + checks in `HasSbit` and `HasWbit` can be enough, or whether we to specialize the flag encoding further?

Do you have any thoughts on the path forward? I notice that a lot of the cases that do not have `s` and `w` bit in 64-bit mode are when `Rex.W` is set (which I originally thought indicates that 'w' bit is not needed).

Comment (anthonycanino) [reply src/coreclr/jit/emitxarch.cpp:160]: @tannergooding  After looking over the instruction manual again, I think I better understand some of the above issue. I think the way the code is written, a lot of the "state" comes from the context in the code, for example, before `HasSBit` is checked https://github.com/dotnet/runtime/blob/bb4e6298227b6a7b7658424cdb105dbe74fa6c1f/src/coreclr/jit/emitxarch.cpp#L12575-L12618

there is a lot of code to determine whether the `acc` form (does not have 's' bit) or the `mi` form should be used.

It would be great to write something like...

```
//------------------------------------------------------------------------
// HasSBit: check if the instruction has the 's' (sign extend) bit.
//
// Arguments:
//    ins - instruction to test
//
// Return Value:
//    true if instruction has the 's' bit, false otherwise
bool emitter::HasSBit(size_t code, instruction ins)
{
    return (code & insCodeMI[ins] != 0 && (CodeGenInterface::instInfo[ins] & INS_FLAGS_Has_Sbit) != 0);
}
```

but that will require larger refactoring, and likewise for the 'w' bit I suspect to truly isolate the logic for 's' and 'w' bit need to be set. Can that be follow up PR changes? Or perhaps there is a simpler encoding to get more of the way there with the current PR?

Comment (tannergooding) [reply src/coreclr/jit/emitxarch.cpp:160]: I'd be fine with that being a follow up PR, but someone else from the codgen team should probably confirm and provide secondary review here (CC. @JulieLeeMSFT)

Comment (anthonycanino) [reply src/coreclr/jit/emitxarch.cpp:160]: Hi @JulieLeeMSFT , is there someone who can provide a secondary review?

Comment (anthonycanino) [reply src/coreclr/jit/emitxarch.cpp:160]: Hi @JulieLeeMSFT @tannergooding , can I get an update on this? I am happy to make further changes, and discuss a follow up PR to keep increasing the CQ of the emitter.

Comment (tannergooding) [reply src/coreclr/jit/emitxarch.cpp:160]: > can I get an update on this? I am happy to make further changes, and discuss a follow up PR to keep increasing the CQ of the emitter.

I expect the delayed response is due to the holiday's and various people being on vacation.

I'm not an owner for any part of the JIT here, just happen to work frequently in it due to owning Hardware Intrinsics, so something from the JIT team will need to also review. As part of that they need to also confirm that handling the oddness around the s and w bits in some future PR is fine. I currently believe it is because the code that the JIT currently uses is already not accounting for this and it appears that the new helper methods preserve the current checks. The biggest "risk" is likely on if the check is now more encompassing and so instructions are tweaked in a way that they shouldn't.

We'll likely need to run some of the stress tests and get PMI diffs to validate no codegen changes, but that will need to happen after the merge conflict is resolved.

CC. @dotnet/jit-contrib

Comment (anthonycanino) [reply src/coreclr/jit/emitxarch.cpp:160]: Thanks @tannergooding . Let me rebase and see how everything works out.

Comment (anthonycanino) [reply src/coreclr/jit/emitxarch.cpp:160]: @tannergooding I believe all checks are good and no asm diffs after rebase.

Can we start some of the fuzzing asm diffs?

Comment (anthonycanino) [reply src/coreclr/jit/emitxarch.cpp:160]: Hi @JulieLeeMSFT @tannergooding , if time permits, can I get some of the stress tests started for this to see if there are asm diffs? Thanks for all the help so far.

**Comment by jakobbotsch** [comment]
The CI asmdiffs leg shows no diffs: https://dev.azure.com/dnceng/public/_build/results?buildId=1499327&view=ms.vss-build-web.run-extensions-tab
Let me run the fuzzers.

/azp run Antigen, Fuzzlyn

**Comment by jakobbotsch** [comment]
The Fuzzlyn failures are known. @kunalspathak, can you look if the Antigen failures are known?

**Comment by kunalspathak** [comment]
> The Fuzzlyn failures are known. @kunalspathak, can you look if the Antigen failures are known?

Yes, all failures are known failures.

File: src/coreclr/jit/emitxarch.cpp
Comment (jakobbotsch) [inline src/coreclr/jit/emitxarch.cpp:11134]: Nits: Seems like the TODO above is outdated with this change, remove it?
Also, I would prefer to not leave author names in comments and to write them in third person or plural first person instead.

File: src/coreclr/jit/emitxarch.cpp
Comment (jakobbotsch) [inline src/coreclr/jit/emitxarch.cpp:11599]: I can't say exactly why these checks are there, but since we currently do not generate cmov I would be fine with removing the checks and leave it up to future work to make sure the emitter supports them if we decide to start generating it again.

File: src/coreclr/jit/emitxarch.cpp
Comment (anthonycanino) [reply src/coreclr/jit/emitxarch.cpp]: Thanks, I have removed the todo and my comment.

File: src/coreclr/jit/emitxarch.cpp
Comment (anthonycanino) [reply src/coreclr/jit/emitxarch.cpp]: Ok, I have gone ahead and removed the comments and the redundant cmov checks.

File: src/coreclr/jit/emitxarch.cpp
Comment (jakobbotsch) [reply src/coreclr/jit/emitxarch.cpp:160]: I had a look at that table in the Intel manual and I am very confused at what these bits are supposed to mean. For example, for the `w` bit we are using it to mean something like "there is a 32-bit instruction form (taking potential operand override prefixes) by setting the lowest bit" if I understand correctly. However the Intel manual describes the w bit with "The value of bit W. in REX has no effect" which I do not understand. In the section in the manual the `w` bit is never part of encodings that has REX.W set, and even so, REX.W is meaningful even when `w = 1`.

Maybe one of you can explain in a bit more detail what is meant by the w/s bits here? It would also be nice to include the explanation in the comments for these methods.

Comment (anthonycanino) [reply src/coreclr/jit/emitxarch.cpp:160]: I think that "w" in the legend for "The value of bit W. in REX has no effect" is actually referring to this, and not the 'w' bit that we have been discussing...

image

because to my understanding, in 64-bit mode, the operands are always 64-bit for the call instruction and the REX.W can be 0 or 1, but it won't change the operand size for this particular instruction.

The "S" in the legend for "If the value of REX.W. is 1, it overrides the presence of 66H" is used like so

image

In other words, those 'S' and 'w' legends are not referring to the 's' and 'w' bits. So your initial understanding of "there is a 32-bit instruction form (taking potential operand override prefixes) by setting the lowest bit" is correct.

Shall I update the comments to make this more clear?

Comment (jakobbotsch) [reply src/coreclr/jit/emitxarch.cpp:160]: Thanks, that explains it.
It would be great if you could expand the comments on `HasWBit` and `HasSBit` to include an explanation and perhaps point to the table in the Intel manual. We could also consider something more descriptive such as `HasRegularWideForm`/`HasRegularSignedForm` with a pointer to this being the w/s bits in the encodings from that section in the Intel manual. But I will leave this up to you -- I don't have a strong opinion as long as it is explained how to interpret the return value.

Comment (anthonycanino) [reply src/coreclr/jit/emitxarch.cpp]: I added some more comments and renamed to HasRegularWideForm. Happy to adjust. Do the comments make it clear what is happening here with respect to the instruction needing the 'w' bit set IF it is in a form where the 'w' bit is present?

I think this is where further refactoring will help per.

Comment (jakobbotsch) [reply src/coreclr/jit/emitxarch.cpp]: I don't think the comments explain much. I would add something like:
```
// Many x86/x64 instructions follow a regular encoding scheme where the
// byte-sized version of an instruction has the lowest bit of the opcode cleared
// while the 32-bit version of the instruction (taking potential prefixes to
// override operand size) has the lowest bit set. This function returns true if
// the instruction follows this format.
// Note that this bit is called `w` in the encoding table in Section B.2 of
// Volume 2 of the Intel Architecture Software Developer Manual.
```
and
```
// As above, many instructions have a regular form used to encode whether the
// instruction takes a sign-extended 1-byte immediate or a sign-extended (when
// in 64-bit) 4-byte immediate, by respectively setting and clearing the second
// lowest bit. This bit is called the s bit in the table in Section B.2.
```
for the other one.

Comment (anthonycanino) [reply src/coreclr/jit/emitxarch.cpp]: Thanks. I've gone ahead and updated the comments. Mostly copied what you pasted into the format in the file. Does it look good?

**Review by jakobbotsch — APPROVED** [review]

**Pull request merged**
